sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedPartitions.scala

```diff
@@ -0,0 +1,253 @@
+  }
+
+  /**
+   * Base optimization support check: the join type is supported.
+   * Note that for some join types(like left outer), whether a certain partition can be optimized
+   * also depends on which side is skewed.
+   */
+  private def supportOptimization(
+      joinType: JoinType,
+      leftStage: QueryStageExec,
+      rightStage: QueryStageExec): Boolean = {
+    val joinTypeSupported = supportedJoinTypes.contains(joinType)
+    val shuffleStageCheck = ShuffleQueryStageExec.isShuffleQueryStageExec(leftStage) &&
+      ShuffleQueryStageExec.isShuffleQueryStageExec(rightStage)
+    joinTypeSupported && shuffleStageCheck
+  }
+
+  private def supportSplitOnLeftPartition(joinType: JoinType) = {
+    joinType == Inner || joinType == Cross || joinType == LeftSemi ||
+      joinType == LeftAnti || joinType == LeftOuter
+  }
+
+  private def supportSplitOnRightPartition(joinType: JoinType) = {
+    joinType == Inner || joinType == Cross || joinType == RightOuter
+  }
+
+  private def getNumMappers(stage: QueryStageExec): Int = {
+    ShuffleQueryStageExec.getShuffleStage(stage).plan.shuffleDependency.rdd.partitions.length
+  }
+
+  def handleSkewJoin(plan: SparkPlan): SparkPlan = plan.transformUp {
+    case smj @ SortMergeJoinExec(leftKeys, rightKeys, joinType, condition,
+    SortExec(_, _, left: QueryStageExec, _),
+    SortExec(_, _, right: QueryStageExec, _))
+      if supportOptimization(joinType, left, right) =>
+      val leftStats = getStatistics(left)
+      val rightStats = getStatistics(right)
+      val numPartitions = leftStats.bytesByPartitionId.length
+
+      val leftMedSize = medianSize(leftStats)
+      val rightMedSize = medianSize(rightStats)
+      logDebug(
+        s"""
+          |Try to optimize skewed join.
+          |Left side partition size: median size: $leftMedSize,
+          | max size: ${leftStats.bytesByPartitionId.max}
+          |Right side partition size: median size: $rightMedSize,
+          | max size: ${rightStats.bytesByPartitionId.max}
+        """.stripMargin)
+
+      val skewedPartitions = mutable.HashSet[Int]()
+      val subJoins = mutable.ArrayBuffer[SparkPlan]()
+      for (partitionId <- 0 until numPartitions) {
+        val isLeftSkew = isSkewed(leftStats, partitionId, leftMedSize)
+        val isRightSkew = isSkewed(rightStats, partitionId, rightMedSize)
+        val leftMapIdStartIndices = if (isLeftSkew && supportSplitOnLeftPartition(joinType)) {
+          getMapStartIndices(left, partitionId, leftMedSize)
+        } else {
+          Array(0)
+        }
+        val rightMapIdStartIndices = if (isRightSkew && supportSplitOnRightPartition(joinType)) {
+          getMapStartIndices(right, partitionId, rightMedSize)
+        } else {
+          Array(0)
+        }
+
+        if (leftMapIdStartIndices.length > 1 || rightMapIdStartIndices.length > 1) {
+          skewedPartitions += partitionId
+          for (i <- 0 until leftMapIdStartIndices.length;
```

Comment:
Do we have a upper limit of the number of `SortMergeJoin`s created here ? I tested the patch and it took a long time to launch the job after optimizing skewed join, the app status was dropped by event bus, the UI was choked due to huge graph.  In my case the number of mappers on both sides were more than 1000.

Reply:
Thanks for you sharing. We will re-design and update later.

Reply:
Thanks. We are back-porting Spark AE features and expect optimizing skewed join to solve many data skew issues we've seen in production.  Please keep me in the loop if any updates.

Reply:
Sure.